Added WSMan Config provider tests

SteveL-MSFT opened the pull request:
added tests for get-item, get-childitem, set-item, new-item remove-item, clear-item, dynamic parameters
merged CoreClr and FullClr code
did not add tests for clientcert handling

Part of https://github.com/PowerShell/PowerShell/issues/4158

Should get line coverage from 50.39% to 72.2%

Lots of error handling code not hit

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

SteveL-MSFT committed:
[Feature]
fixed issues in WSMan Config Provider
added tests for get-item, get-childitem, set-item, new-item remove-item, clear-item, dynamic parameters
merged CoreClr and FullClr code

iSazonov commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+            }
+            finally {
+                Remove-Item -Path "WSMan:\localhost\Listener\$listenerName" -Force
+                $newListener.PSPath | Should Not Exist
```
Why the Should in finally block? Maybe add a comment?

SteveL-MSFT replied:
The tests are for `new-item` and corresponding `remove-item`.  `Should` validates that the `remove-item` worked.  The test title indicates this and I thought was obvious enough.  In case some other `Should` validation fails before the `remove-item` I still want to cleanup.  Do you think a comment is needed?

iSazonov replied:
Thanks for clarify. 

Closed.

iSazonov commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+        }
+    }
+
+    Context "Get-Help tests" {
```
Should we move the test to a file with all Help tests? Tum more what we want to move Help subsystem in separate module.

SteveL-MSFT replied:
This test is a bit different from the HelpSystem.Tests.ps1 as it's testing the implementation of ICmdletProviderSupportsHelp in this provider (not testing the HelpSystem).  But more than that, it's specific to Windows and requires elevation.

iSazonov replied:
Thanks for clarify. I think we should mention this in a comment or `Context` description.

mirichmo commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+
+        It "New-Item and Remove-Item for a plugin using a file" {
+            $fileXml = @"
+<PlugInConfiguration Name="TestPlugin2" Filename="%windir%\system32\pwrshplugin.dll" SDKVersion="2" XmlRenderingType="text" Enabled="false"
```
This is the Windows PowerShell plugin path. It matches the config you have been using throughout, but it will need to be updated

SteveL-MSFT replied:
I'll update it to the one for PSCore6

mirichmo commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+        It "New-Item and Remove-Item for an initialization parameter" {
+            try {
+                $parameter = New-Item -Path WSMan:\localhost\Plugin\TestPlugin\InitializationParameters `
+                    -ParamName foo -ParamValue bar
```
We should not use "foo" and "bar" since they violate PolyCheck. Please use other string values.

SteveL-MSFT replied:
I believe policheck allows it for code, but not in strings given to users.  In any case, if we need to fix this, we should have a separate PR for all the code.

mirichmo commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+                (Get-ChildItem $securityObj.PSPath | Where-Object { $_.Name -eq 'sddl' }).Value | Should Be $sddl
+            }
+            finally {
+                Remove-Item $security.PSPath
```
Does Remove-Item on a non-existant path throw an error?

SteveL-MSFT replied:
I'm testing Remove-Item and also trying to cleanup.  I think it may be better since @iSazonov gave similar feedback and have some redundancy where I do the `remove-item` test within the `try` and still do cleanup.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+
+        It "WSMan Config Provider starts WinRM if it is stopped" {
+            try {
+                Stop-Service WinRM
```
`Stop-Service` should be waiting by default, it has a `-nowait` switch

iSazonov replied:
Closed.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+        }
+    }
+
+    Context "Tab complete tests" {
```
I'll move them

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+
+        if ($IsWindows) {
+            $badCredentialError = 1326
+            $pluginXml = [xml](winrm g winrm/config/plugin?name=microsoft.powershell -format:xml)
```
I'll make the change

SteveL-MSFT replied:
Had to change it back as using Install-PowerShellRemoting or Enable-PSRemoting was making things complicated even though my tests had no dependency on remoting actually working.  Since I'm just relying on a well known existing plugin configuration, I think using microsoft.powershell seems reasonable.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+            $testPluginPath = "WSMan:\localhost\Plugin\TestPlugin"
+            $testUser = (Get-Random)
+            $testPass = "Secret123!"
+            $null = net user $testUser $testPass /add
```
Not needed, I don't do any actual remoting.  I just need a valid local account to set configuration (runas)

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+
+    Context "Misc tests" {
+        It "Can remove and add wsman drive" {
+            $w = Get-PSDrive -Name WSMan
```
will change

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+            @{path = "localhost\plugin"; parameter = "-ru"; expected = "RunAsCredential"},
+            @{path = "localhost\plugin"; parameter = "-us"; expected = "UseSharedProcess"},
+            @{path = "localhost\plugin"; parameter = "-au"; expected = "AutoRestart"},
+            @{path = "localhost\plugin"; parameter = "-pr"; expected = "ProcessIdleTimeoutSec"},
```
There should already be tabcomplete tests to cover that.  This is specifically testing WSMan Config Provider's returning dynamic parameters

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+            # https://github.com/PowerShell/PowerShell/issues/4744
+            # TODO: move to test cases above once working
+        }
+
```
will fix

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+
+        It "Set-Item on plugin Security Resource '<property>' property with '<value>' should succeed" -TestCases @(
+            @{property="Sddl"; value="O:NSG:BAD:P(A;;GA;;;BA)"},
+            @{property="SDDL"; value="O:NSG:BAD:P(A;;GA;;;BA)(A;;GA;;;IU)(A;;GA;;;RM)S:P(AU;FA;GA;;;WD)(AU;SA;GXGW;;;WD)"}
```
I'll add a comment, but the second one is the default SDDL, the first is a truncated version

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,478 @@
+        ) {
+            param($property, $value)
+            Set-Item "$testPluginPath\InitializationParameters\$property" $value -WarningAction SilentlyContinue
+            (Get-Item "$testPluginPath\InitializationParameters\$property").Value | Should Be $value
```
I thought about it while writing the tests, but there's enough differences in the properties and validation I think it would be overly complicated to combine them

SteveL-MSFT committed:
[feature]
address PR feedback

SteveL-MSFT commented:
CI failure waiting on https://github.com/PowerShell/PowerShell/pull/4802 to be merged

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+    }
+
+    AfterAll {
+        $PSDefaultParameterValues = $originalDefaultParameterValues
```
Set $Global:PSDefaultParameterValues instead

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+
+    Function Test-Plugin($plugin, $expectedMissingProperties, $expectedMissingAttributes) {
+        $plugin.PSPath | Should Exist
+        $plugin | Should Not BeNullOrEmpty
```
This check seems unnecessary as $plugin.PSPath is already checked.

SteveL-MSFT replied:
true, will remove

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+        $plugin | Should Not BeNullOrEmpty
+        $testPluginXml = [xml](winrm g winrm/config/plugin?name=$($plugin.Name) -format:xml)
+        $pluginProperties = Get-ChildItem $plugin.PSPath
+        $pluginProperties.Count | Should Be (($testPluginXml.PluginConfiguration | Get-Member -Type Properties).Count + $expectedMissingProperties.Count - $expectedMissingAttributes.Count)
```
Please move the value calculated for Should Be to a variable. It will make it more readable.

SteveL-MSFT replied:
ok

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+            { Get-PSDrive -Name wsman -ErrorAction Stop } | ShouldBeErrorId "GetLocationNoMatchingDrive,Microsoft.PowerShell.Commands.GetPSDriveCommand"
+            $wsmanDrive2 = $wsmanDrive | New-PSDrive -PSProvider WSMan
+            $wsmanDrive2 | Should BeOfType System.Management.Automation.PSDriveInfo
+            $wsmanDrive2.Name | Should Be "WSMan"
```
Should this be BeExactly

SteveL-MSFT replied:
will change

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+            $wsmanDrive2 = $wsmanDrive | New-PSDrive -PSProvider WSMan
+            $wsmanDrive2 | Should BeOfType System.Management.Automation.PSDriveInfo
+            $wsmanDrive2.Name | Should Be "WSMan"
+            $wsmanDrive2.Provider.Name | Should Be "WSMan"
```
Should this be BeExactly

SteveL-MSFT replied:
will change

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+            param($property, $value, $expected)
+            $resource = Get-ChildItem "$testPluginPath\Resources" | Select-Object -First 1
+            Set-Item "$($resource.PSPath)\$property" $value -WarningAction SilentlyContinue
+            (Get-Item "$($resource.PSPath)\$property").Value | Should Be $expected
```
Should this be BeExactly

SteveL-MSFT replied:
will change

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+            }
+            finally {
+                if (Test-Path "WSMan:\localhost\Plugin\TestPlugin2\") {
+                    Remove-Item "WSMan:\localhost\Plugin\TestPlugin2\" -Force
```
This can be ```Remove-Item "WSMan:\localhost\Plugin\TestPlugin2\" -Force -ErrorAction SilentlyContinue``` and remove the ```Test-Path```.

SteveL-MSFT replied:
true, will change

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+            }
+            finally {
+                if (Test-Path $resource.PSPath) {
+                    Remove-Item $resource.PSPath -Force
```
Same as above.

SteveL-MSFT replied:
will change

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+            }
+            finally {
+                if (Test-Path $parameter.PSPath) {
+                    Remove-Item $parameter.PSPath -Force
```
Same as above.

SteveL-MSFT replied:
will change

adityapatwardhan commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,441 @@
+            }
+            finally {
+                if (Test-Path $security.PSPath) {
+                    Remove-Item $security.PSPath -Force
```
Same as above.

SteveL-MSFT replied:
will change

SteveL-MSFT commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -1006,3 +1006,65 @@ Describe "Tab completion tests with remote Runspace" -Tags Feature {
+    }
+
+    AfterAll {
+        $PSDefaultParameterValues = $originalDefaultParameterValues
```
Why would we want to mess with global scope?

iSazonov replied:
:-) I asked @JamesWTruher previously the same and he said that is good to exclude scoped isssues.

adityapatwardhan replied:
@JamesWTruher could you point to the PR where we had problems with scoping and PSDefaultParameterValues

SteveL-MSFT replied:
Actually, had an issue with scoping where it affected the CIM cmdlet tests on non-Windows.  Changed it to global and it fixed the issue.  Would like to understand what is happening here as it's not intuitive.

SteveL-MSFT commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -1006,3 +1006,65 @@ Describe "Tab completion tests with remote Runspace" -Tags Feature {
+    It "Tab completion works correctly for Listeners" {
+        $path = "wsman:\localhost\listener\listener"
+        $res = TabExpansion2 -inputScript $path -cursorColumn $path.Length
+        $l = Get-ChildItem WSMan:\localhost\Listener
```
will fix

SteveL-MSFT commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,433 @@
+
+        if ($IsWindows) {
+            $badCredentialError = 1326
+            $endpoint = & "$PSHOME\Install-PowerShellRemoting.ps1"
```
I don't need remoting to actual work for the purpose of these tests, I think I'll simplify it by just getting the first endpoint and validating against that

mirichmo commented:
@SteveL-MSFT Looks like a few tests are failing

SteveL-MSFT commented:
@mirichmo I'm going to revert back to using microsoft.powershell plugin as my base plugin since that seems to be well defined and trying to make it work with enable-psremoting turned out to be more complicated than it's worth particularly since I'm not actually using remoting for these tests.

SteveL-MSFT committed:
[feature]
revert to using microsoft.powershell plugin as base for test
keeps it simple rather than relying on additional scripts or cmdlets since these tests don't require remoting

SteveL-MSFT committed:
[feature]
address Aditya's feedback

SteveL-MSFT commented:
@adityapatwardhan your feedback has been addressed

iSazonov commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,436 @@
+        $Global:PSDefaultParameterValues = $originalDefaultParameterValues
+        if ($IsWindows) {
+            $null = winrm d winrm/config/plugin?Name=TestPlugin
+            $null = net user $testUser /DELETE
```
Maybe use as in line 50 `> $null`?

SteveL-MSFT replied:
I'll change it to `$null =` for consistency

iSazonov commented on test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1:
```diff
@@ -0,0 +1,436 @@
+            $securityContainer = Get-ChildItem "$pluginPath\Resources\$($resource.Name)\Security" | Select-Object -First 1
+            $securityProperties = Get-ChildItem $securityContainer.PSPath
+            $skippedAttributes = @("ParentResourceUri","xmlns") # these are added by the WSMan Config Provider but not in the original xml
+            # ParentResourceUri is added by the Config Provider, xmlns existing seems to be dependent on WinRM which is inconsistent
```
Please place a comment on new line.

mirichmo replied:
I think the request here was to move the comment on line 115 to a new line, not insert a newline character after it.

This isn't a blocking issue for the PR. I'm just clarifying intent.

iSazonov reviewed (commented):
LGTM with monor comments.

adityapatwardhan reviewed (approved):
LGTM

SteveL-MSFT committed:
[feature]
address some cosmetic feedback

mirichmo commented:
Thanks for all the updates @SteveL-MSFT. I'll merge the PR once the tests pass

SteveL-MSFT commented:
Maybe I shouldn't have rebased, picked up https://github.com/PowerShell/PowerShell/pull/4811 which is causing some failures.  I'll look into it.

SteveL-MSFT committed:
[feature]
based on discussion with Jim, we should skip resx checks on non-Windows

SteveL-MSFT commented:
Travis-CI failures due to resx tests not finding resources that don't exist there (like cimcmdlets).  After discussion with @JamesWTruher we decided that it is ok to skip those tests on non-Windows.  Easier to just make that change in this PR than making another PR and pulling it in.
